ec_tools/semi_integration.py

```diff
@@ -1,63 +1,456 @@
+
+
+
+Fast Riemann
+^^^^^^^^^^^^
+The following algorithm was introduced by Pajkossy et al. in 1984 [2] and is based on the Riemann-Liouville transformation (RLT). 
+Its big advantage is, that the computation time increases only linearly with the number of points (N). 
+Here it is necessary to define some input variables (beside the t & I(t) data), where q is equal to v, describing a semi integration or semi differentiation. 
+:math:`\Delta_t` defines the constant time intervall (i.e. :math:`t_2 - t_1`) and :math:`c_1, c_2` are constant values,
+which we set by default to :math:`c_1=8, c_2=2` (but still changeable). With these variables, the procedure of the algorithm can be described as pseudo code:
+
+**Input** 
+:math:`q, N, \Delta_t, c_1, c_2, I`
+
+:math:`t_0 = \Delta_t N^{1/2}` 
+
+:math:`a_0 = \sin(\pi q)/(\pi qt_0^q)`
+
+**For** :math:`i = 0,2c_1c_2`
+
+    :math:`j = i-c_1c_2`
+    
+    :math:`a_j = (a_0/c_2)\exp(j/c_2)`
+
+    :math:`t_j = t_0 \exp(-j/qc_2)`
+
+    :math:`w_1(i) = t_j / (\Delta_t + t_j)`
+
+    :math:`w_2(i) = a_j(1-w_1(i))`
+
+    :math:`s(i) = 0`
+
+**For** :math:`k=1,N`
+
+    :math:`R(k)=0`
+    
+    **For** :math:`i=0,2c_1c_2`
+    
+        :math:`s(i)= s(i)w_1(i) + I(k)w_2(i)`
+
+        :math:`R(k) = R(k) + s(i)`
+
+Here, :math:`R` represents the calculated semi integral, i.e. :math:`R \approx \frac{d^{v}}{dt^{v}} I(t)`.
+
 """
 
 import numpy as np
+from transonic import jit
 
 
-def prepare_kernel(q, delta_x, N, c1, c2):
+def semi_integration(
+    y, x, v=-0.5, alg="frlt", transonic_backend="transonic", d_tol=1e-5
+):
     r"""
-    Setup the integration kernel with the order q, the x interval delat_x, the length of the array N,
-    and the filter constants c1 and c2.
+    To perform a semi integration (``v`` =-0.5) or semi differentiation (``v`` =0.5) 
+    on given data (``y`` and ``t`` ) with speed up by transonic (with numba or pythran backend) or without simply using python
+    different methods are implemented.
+
+    Available algorithms (``alg`` ):
+
+    ``frlt``: Fast Test Riemann-Liouville transformation
+
+    ``g1``: Gruenwald
+
+    ``r1``: Riemann and Liouville
+
+    Available settings (``transonic_backend`` ):
+
+    ``python``: Python implementation
+
+    ``numba``: Numba package with jit(just in time)
+
+    ``pythran``: Transonic package with numba backend
+
+    """
+
+    # Calc average step size
+    deltas = np.diff(x)
+    delta_x = deltas.mean()
+
+    # warning if avg. delta_x differs to much from single ones
+    if np.max(np.abs(deltas/delta_x-1)) >= d_tol:
+        print(
+            "Warning: step size tolerance reached!\nAverage stepsize differs up to ", np.max(np.abs(deltas/delta_x-1))>= d_tol
+        )
+
+    if alg not in ["g1", "r1", "frlt"] or transonic_backend not in [
+        "python",
+        "numba",
+        "pythran",
+    ]:
+        raise ValueError("\nNo matching setting for alg or transonic_backend")
+
```

Comment:
```suggestion
# Note: It is not possible to simply pass the backend variable to transonic. It is therefore set manually.
# Maybe it could work with cases.
```